### ISSUE TITLE
Privacy: Prohibit access/use of sensitive data (e.g. location) by CDMs

### ISSUE THREAD

**Issue by ddorwin** [w3c/encrypted-media#158]
In https://github.com/w3c/encrypted-media/issues/157#issuecomment-208844577, @mwatson2 says:
>For online viewing, services may indeed apply geographic restrictions. ...it is a server function to apply these restrictions, not something that is done by the DRM. This is important to recognize because there would be privacy implications if the CDM could access your location.

While we assume the CDM cannot access or use the client's/user's location, I'm not sure it is currently expressly prohibited by the spec.

More generally, the CDM should not use (have access to?) or expose data that is not generally available to web applications or is generally protected by a user permission and/or prompt. Location is a primary example, but there are others, both exposed to the web (i.e. user media, such as camera and mic) and not (i.e. LAN details or devices).

While the examples above may seem clear cut, the phrasing could be tricky, especially since unsandboxed CDMs often do have such access and some CDMs use, for example, Distinctive Identifiers not otherwise exposed.

Note that preventing exposure of such data is not sufficient since even use of them could allow them to be derived (i.e. via a series of licenses).

**Comment by mwatson2** [comment]
In my statement I was assuming that the text around privacy and distinctive identifiers excluded the exposure by the CDM of personally identifiable information generally, but perhaps we need to review tighten the text there.

In general, when we talk about restricting "exposure" of information we should include within this "indirect exposure" (i.e. ability to derive information over time). CDMs may "use" various kinds of information, for example to make the identifiers they use distinctive ( = not shared by other users / devices), but the important point is that any exposed information can be reset to prevent tracking over time. Put another way, no time-persistent / non-resettable information should be exposed (directly or indirectly).

**Comment by josephlhall** [comment]
Hi, Joe Hall from W3C's Privacy Interest Group (PING) here (also a lurker in EME/public-html-media).

It sounds like EME calls and even CDMs don't need to request permissions through the browser for access to certain sensitive data like those called out in the [Secure Contexts](https://w3c.github.io/webappsec-secure-contexts/#threat-risks) draft?

PING would be happy to discuss this further, although it sounds like from what @mwatson2 says near the end there are limited privacy implications.

**Comment by steelejoe** [comment]
@josephlhall as @mwatson2 said the only piece of sensitive information which has been called out as necessary so far is the [Distinctive Identifier](https://w3c.github.io/encrypted-media/#definitions). We do have [language around user consent](https://w3c.github.io/encrypted-media/#get-distinctive-identifier-consent-status) there, although it is still being worked on. 

@ddorwin As far as other types of sensitive information, I am not opposed to adding geo-location. I am not sure whether a black list approach will be effective though, as more stuff will arise in the future. E.g. should the CDM have microphone access. Can we tie this back to [Secure Contexts](https://w3c.github.io/webappsec-secure-contexts/#threat-risks) as Joseph is suggesting? Since that is a draft I am not sure how that works.

**Comment by ddorwin** [comment]
Hi @josephlhall! Thanks for the pointer. We have an open [Issue](https://w3c.github.io/encrypted-media/#h-issue2) in the spec regarding review of the security and privacy sections. It would be great to have more input here, though you may want to hold off on a full review until we address some related issues (mostly #117 but also #110, #96, and #101) since the text could change. Comments in those issues are also welcome.

@steelejoe, I did not mean to imply we should have a blacklist - these were only examples. Whatever language we settle on should cover other such items and be future-proof.

Secure Contexts' [risks](https://w3c.github.io/webappsec-secure-contexts/#threat-risks) section looks like a good start. In addition to referencing it, we might want to specifically highlight some items (e.g. location). We may also want to provide some additional examples of things that are covered by the statement but may not be obvious given the examples. For example, the first item covers a user's OS account name and even directory paths, which can contain private information, but this may not be obvious.

There may also be the opposite problem of preventing things that are allowed by EME (given permission): The fourth item is a reference loop to EME and could in theory prevent Distinctive Identifiers. The fifth item may be similarly problematic for persistent state.

I wonder if there should be a more general document about things user agents should avoid exposing. Secure Contexts could even reference it as "any API allowing access to such things should require a secure context." @josephlhall, any thoughts on this?

<br/>@steelejoe, I'm not sure what "deriving the identifier" refers to. My use of "derived" should perhaps have been "inferred." As an example, if a CDM has access to the user's location, it could support licenses with region restrictions. Then, a website could determine the user's location to the granularity of the "regions" by issuing multiple licenses/keys, each with a different set of region restrictions, then checking which keys are "usable."

Without a sandbox, it may not be possible (for the user agent) to prevent access to, for example, the GPS location, but we can at least make it clear that it is important that such values and functionality  MUST NOT be used. It may then be up to the user agent to ensure this as part of the vetting process before integration.

**Comment by steelejoe** [comment]
@ddorwin I was referring to this line:

> Note that preventing exposure of such data is not sufficient since even use of them could allow them to be **derived **(i.e. via a series of licenses).

But your response makes sense to me now. I agree about referencing Secure Context risks section. That seems useful.

**Comment by paulbrucecotton** [comment]
@ddorwin, @mwatson2, @jdsmith3000 :  We need to start making progress on this V1 issue.  Can one of the Editors self-assign it and start making a detailed proposal or PR based on the items listed in David's comment<https://github.com/w3c/encrypted-media/issues/158#issuecomment-209669987>?

**Comment by mwatson2** [comment]
I'm not sure we need to normatively reference specific issues. I'm wondering if we can establish a normative principle that would hold for both current and future privacy risks ?

The principle is that the EME API MUST NOT expose* information that is not otherwise available to the page, _except as explicitly described in the EME specification_ (e.g. Distinctive Identifiers, audio / video capabilities, output protection, persistent store). Put another way, it must not change the attack surface in unspecified ways.

* here we must find a different word or have a sufficiently wide definition of "expose" to include the kind of inferences @ddorwin described.

Note that _how_ the CDM is prevented from exposing such information is not in scope for us. Sandboxing is one approach. For those cases where the CDM implementor and UA implementor are one and the same, then they can simple design the CDM such that it does not expose such information.

**Comment by ddorwin** [comment]
@mwatson2's suggested principle-based approach sounds like a good start and is more future-proof than a blacklist. However, I think we should follow it with explicit examples of known concerns while still noting this requirement is not limited to these.

For example:
>The type of information covered by this requirement includes but is not limited to:
* Location, including geolocation
* Credentials or identifiers other than Distinctive Identifiers (we can update this after #117 is fixed)
* OS account name and other potential PII
* Local directory paths, which may contain similar information.
* Internal/private network details (hopefully there is a term for this we can use)
* Local devices, including but not limited to Bluetooth, USB, and user media.
* User state not associated with or stored as a result of these APIs.

We could then perhaps say, "See also Secure Contexts' [risks section](https://w3c.github.io/webappsec-secure-contexts/#threat-risks)."

Regarding the actual text of the principle:
* "otherwise available to the page" might be problematic since geolocation is available to the page, but requires a permission. Perhaps we need to add "without user permission/consent" or something, though that depends on the requirements of individual specs and whether the implementation conforms to those specs.
* We could use "expose or enable the inference or derivation of" in place of "expose."

**Comment by mwatson2** [comment]
> "otherwise available to the page" might be problematic since geolocation is available to the page, but requires a permission.

I interpret something which requires permission, but for which permission has not been granted, as "unavailable" to the page.

Do we wish to make a blanket requirement that only information already available to the page without user permission be allowed to be exposed by the CDM, or do we allow the CDM to expose information which requires permission and for which permission has been granted ?

**Comment by mwatson2** [comment]
I created a proposal for this issue [here](https://github.com/w3c/encrypted-media/compare/gh-pages...mwatson2:issue-158), but for some reason GitHub will not allow me to create a pull request.

**Comment by mwatson2** [comment]
Ok, it seemed GitHub needed some time to think about it. The PR is [here](https://github.com/w3c/encrypted-media/pull/199).

**Comment by mwatson2** [comment]
Reviewed at editors' meeting (@mwatson2 @jdsmith3000). We need to settle the question of whether this restriction applies to information which is available to the page with user permission (assuming the user has given permission). @ddorwin ?

**Comment by ddorwin** [comment]
I think we should start with "no." This can always be relaxed in the future if there is a need, but a) this could be quite complex to normatively/algorithmically describe or at least take some flushing out to make sure we don't miss corner cases, and b) there may be concerns with essentially adding attestation for something like geolocation.

**Comment by mwatson2** [comment]
@ddorwin @mwatson2 @jdsmith3000 reviewed this and agreed to merge with changes as noted in line notes.
